XenAdmin/Controls/USB/UsbList.cs

```diff
@@ -0,0 +1,416 @@
+                    ((UsbRow)row).UpdateDetails();
+            }
+        }
+
+        private void buttonUsage_Click(object sender, EventArgs e)
+        {
+            UsbUsageDialog dialog = new UsbUsageDialog(selectedUsb);
+            dialog.ShowDialog(Program.MainWindow);
+        }
+
+        private void buttonAttach_Click(object sender, EventArgs e)
+        {
+            if (_xenObject is VM)
+            {
+                AttachUsbDialog dialog = new AttachUsbDialog(_xenObject as VM);
+                dialog.ShowDialog(Program.MainWindow);
+            }
+        }
+
+        private void buttonDetach_Click(object sender, EventArgs e)
+        {
+
+        }
+
+        private void UsbGridView_SelectionChanged(object sender, EventArgs e)
+        {
+            selectedUsb = null;
+
+            try
+            {
+                if (UsbGridView.Rows.Count <= 0)
+                {
+                    buttonUsage.Enabled = false;
+                }
+                else
+                {
+                    HostUsbRow row = (HostUsbRow)UsbGridView.SelectedRows[0];
+                    selectedUsb = row.pusb;
+                    buttonUsage.Enabled = true;
+                }
+            }
+            catch
+            {
+                buttonUsage.Enabled = false;
+            }
+            finally
+            {
+                if (null != selectedUsb && selectedUsb.passthrough_enabled)
+                    buttonUsage.Text = Messages.USBLIST_DISABLE_PASSTHROUGH;
+                else
+                    buttonUsage.Text = Messages.USBLIST_ENABLE_PASSTHROUGH;
+            }
+        }
+
+        class UsbRow : DataGridViewExRow
+        {
+            private IXenObject _xenObject;
+
+            public UsbRow(IXenObject xo)
+            {
+                _xenObject = xo;
+                xo.PropertyChanged += Server_PropertyChanged;
+            }
+
+            public virtual void UpdateDetails()
+            {
+
+            }
+
+            public void DeregisterEvents()
+            {
+                _xenObject.PropertyChanged -= Server_PropertyChanged;
+            }
+
+            private void Server_PropertyChanged(object sender, PropertyChangedEventArgs e)
+            {
+                Program.Invoke(Program.MainWindow, UpdateDetails);
+            }
+        }
+
+        class HostUsbRow : UsbRow
+        {
+            private PUSB _pusb;
+
+            private DataGridViewTextBoxCell locationCell = new DataGridViewTextBoxCell();
+            private DataGridViewTextBoxCell descriptionCell = new DataGridViewTextBoxCell();
+            private DataGridViewTextBoxCell usageCell = new DataGridViewTextBoxCell();
+            private DataGridViewTextBoxCell vmCell = new DataGridViewTextBoxCell();
+
+            public PUSB pusb
+            {
+                get { return _pusb; }
+            }
+
+            public HostUsbRow(PUSB pusb): base(pusb)
+            {
+                _pusb = pusb;

```

Comment:
_xenObject in the base class already contains the pusb, so I don't think we need the `_psub ` field here at all.

Reply:
Will fix.